libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,305 @@
+		return;
+
+	smtutil::Expression condition = encodeExpression(*_if.condition);
+	m_solver->push();
+	m_solver->addAssertion(condition == constantValue(0));
+	CheckResult result = m_solver->check({}).first;
+	m_solver->pop();
+	if (result == CheckResult::UNSATISFIABLE)
+	{
+		Literal trueCondition = m_dialect.trueLiteral();
+		trueCondition.location = locationOf(*_if.condition);
+		_if.condition = make_unique<yul::Expression>(move(trueCondition));
+	}
+
+	m_solver->push();
+	m_solver->addAssertion(condition != constantValue(0));
+	CheckResult result2 = m_solver->check({}).first;
+	m_solver->pop();
+	if (result2 == CheckResult::UNSATISFIABLE)
+	{
+		Literal falseCondition = m_dialect.zeroLiteralForType(m_dialect.boolType);
+		falseCondition.location = locationOf(*_if.condition);
+		_if.condition = make_unique<yul::Expression>(move(falseCondition));
+		_if.body = yul::Block{};
+	}
+
+	m_solver->push();
+	m_solver->addAssertion(condition != constantValue(0));
+
+	ASTModifier::operator()(_if.body);
+
+	m_solver->pop();
+}
+
+ReasoningBasedSimplifier::ReasoningBasedSimplifier(
+	Dialect const& _dialect,
+	set<YulString> const& _ssaVariables
+):
+	m_dialect(_dialect),
+	m_ssaVariables(_ssaVariables),
+	m_solver(make_unique<smtutil::SMTPortfolio>())
+{
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeExpression(yul::Expression const& _expression)
+{
+	return std::visit(GenericVisitor{
+		[&](FunctionCall const& _functionCall)
+		{
+			if (auto const* dialect = dynamic_cast<EVMDialect const*>(&m_dialect))
+				if (auto const* builtin = dialect->builtin(_functionCall.functionName.name))
+					if (builtin->instruction)
+						return encodeEVMBuiltin(*builtin->instruction, _functionCall.arguments);
+			return newRestrictedVariable();
+		},
+		[&](Identifier const& _identifier)
+		{
+			if (
+				m_ssaVariables.count(_identifier.name) &&
+				m_variables.count(_identifier.name)
+			)
+				return m_variables.at(_identifier.name);
+			else
+				return newRestrictedVariable();
+		},
+		[&](Literal const& _literal)
+		{
+			return literalValue(_literal);
+		}
+	}, _expression);
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeEVMBuiltin(
+	evmasm::Instruction _instruction,
+	vector<yul::Expression> const& _arguments
+)
+{
+	vector<smtutil::Expression> arguments = applyMap(
+		_arguments,
+		[this](yul::Expression const& _expr) { return encodeExpression(_expr); }
+	);
+	switch (_instruction)
+	{
+	case evmasm::Instruction::ADD:
+		return wrap(arguments.at(0) + arguments.at(1));
+	case evmasm::Instruction::MUL:
+		return wrap(arguments.at(0) * arguments.at(1));
+	case evmasm::Instruction::SUB:
+		return wrap(arguments.at(0) - arguments.at(1));
+	case evmasm::Instruction::DIV:
+		return smtutil::Expression::ite(
+			arguments.at(1) == constantValue(0),
+			constantValue(0),
+			arguments.at(0) / arguments.at(1)
+		);
+	case evmasm::Instruction::SDIV:
+		// TODO - rounding is different in SMT than in EVM
```

Comment:
We could use
```suggestion
		smtutil::Expression::ite(
			arguments.at(1) == constantValue(0),
			constantValue(0),
			bv2int(int2bv(arguments.at(0)) / int2bv(arguments.at(1)))
		);
```
for now (I checked - we use signed bitvector division like that in ``test/formal`` and it does round correctly) - but probably not much sense in it, because it ever occurring will probably immediately overwhelm the solvers...
Otherwise it'll probably have to be deeply nested ``ite``s depending on the signs...
```suggestion
		return smtutil::Expression::ite(
			arguments.at(1) == constantValue(0),
			constantValue(0),
			smtutil::Expression::ite(
				arguments.at(0) >= 0,
				smtutil::Expression::ite(
					arguments.at(1) >= 0,
					// >= 0, >= 0
					arguments.at(0) / arguments.at(1),
					// >= 0, < 0
					-(arguments.at(0) / -arguments.at(1))
				),
				smtutil::Expression::ite(
					arguments.at(1) >= 0,
					// < 0, >= 0
					-(-arguments.at(0) / arguments.at(1)),
					// < 0, < 0 (can overflow!)
					wrap(-arguments.at(0) / -arguments.at(1)) // could replace wrapping by an ite(arguments.at(1) == -1, ...
				)
		);
```

But maybe too much for this PR and better to just keep it unimplemented...

Reply:
Did you check the helper in `SMTEncoder::division`? We could pretty much use the same thing.

Reply:
> I actually don't think we need wrap in the last case.

We need either wrap or a special case for ``/ -1``, don't we?
``-2^255 / -1`` is ``2^255``, but that's ``> 2^255 - 1`` and wraps around to ``-2^255`` again, doesn't it?

Actually ``wrap`` probably won't do the right thing, but we need to handle the "division by minus one overflow" somehow, don't we?

Reply:
Yea I think so too

Reply:
It's only those specific two values ``arguments.at(0) == -2^255`` and ``arguments.at(1) == -1``, though - nothing else needs wrapping.